llvm/lib/IR/AutoUpgrade.cpp

```diff
@@ -2550,6 +2554,11 @@ static Value *upgradeNVVMIntrinsicCall(StringRef Name, CallBase *CI,
     Intrinsic::ID IID = (Name == "fabs.ftz.f") ? Intrinsic::nvvm_fabs_ftz
                                                : Intrinsic::nvvm_fabs;
     Rep = Builder.CreateUnaryIntrinsic(IID, CI->getArgOperand(0));
+  } else if (Name.consume_front("ex2.approx.")) {
+    // nvvm.ex2.approx.{f,ftz.f,d,f16x2}
+    Intrinsic::ID IID = Name.starts_with("ftz") ? Intrinsic::nvvm_ex2_approx_ftz
+                                                : Intrinsic::nvvm_ex2_approx;
```

Comment:
Would it be out of scope for AutoUpgrade to check the FP flags on the surrounding function and if they agree with the FTZ modifier, switch to using `@llvm.exp2`?

Reply:
This seems like too complex and potentially problematic a transformation to make in AutoUpgrade. We do similar transformations in `convertNvvmIntrinsicToLlvm` which is called during InstCombine, but I think we don't do this for "approx" intrinsics because the semantics don't match the llvm versions.

Reply:
For our target I think we always lower exp2() to ex2.approx so I think it may be worth having a conversion in InstCombine.